bson/src/main/org/bson/codecs/pojo/AutomaticPojoCodec.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.pojo;
+
+import org.bson.BsonReader;
+import org.bson.BsonWriter;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import static java.lang.String.format;
+
+final class AutomaticPojoCodec<T> extends PojoCodec<T> {
```

Comment:
Argh, I messed up when testing - its needed to wrap the PojoCodec provided by AutomaticCodecProvider - so to provide meaningful error messages if encoding or decoding fails.